test/test_dataset_mutability.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Third-party imports
+import numpy as np
+import pandas as pd
+from copy import deepcopy
+
+# First-party imports
+from gluonts.dataset.common import ListDataset
+from gluonts.dataset.field_names import FieldName
+from gluonts.transform import (
+    AddObservedValuesIndicator,
+    Chain,
+    AddAgeFeature,
+    AddConstFeature,
+)
+
+
+ds1 = [
+    {
+        "start": pd.Timestamp("2020/01/01", freq="1D"),
+        "target": np.array(
+            [1, 2, 3, np.nan, 5, np.nan, 7, np.nan, np.nan, 10]
+        ),
+    }
+]
+ds2 = ListDataset(
+    [
+        {
+            "start": "2020/01/01",
+            "target": [1, 2, 3, np.nan, 5, np.nan, 7, np.nan, np.nan, 10],
+        }
+    ],
+    freq="1D",
+)
+
+ds1_c = deepcopy(ds1)
+ds2_c = deepcopy(ds2)
+
+transform = Chain(
+    [
+        AddObservedValuesIndicator(
+            target_field=FieldName.TARGET,
+            output_field=FieldName.OBSERVED_VALUES,
+        ),
+        AddAgeFeature(
+            target_field=FieldName.TARGET,
+            output_field="age_feature",
+            pred_length=1,
+        ),
+        AddConstFeature(
+            target_field=FieldName.TARGET,
+            output_field="constant",
+            pred_length=1,
+        ),
+    ]
+)
+
+# test that using twice the transformation gives the same result
+for d in [ds1, ds2]:
+    out1 = list(transform(d, is_train=True))
+    out2 = list(transform(d, is_train=True))
```

Comment:
Yes, I did this initially because I was checking some fields of the output after going through it and since it is a generator it was empty. With the current version this is not needed.